### PR TITLE
Add go module support

### PR DESCRIPTION
**Pull request by saschagrunert**
This PR adds a `go.mod` and `go.sum` to the project which contains the latest dependencies. It would be awesome if we could tag a release according to semver after this PR got merged, WDYT? :innocent:

### CODE REVIEW

**Comment by lucab** [comment]
We touched this topic in the past with @arithx, and we are a bit cautious.

Relevant sharp edges:
 * there is no overall `go-systemd` package, just individual sub-packages. Versioning is shared, though.
 * this repository had tagged releases since the beginning, but gomod doesn't understand the format. So the starting version likely has to be with `v20.0.0` (next tag).
 * the API re-exports types from other libraries (e.g. godbus). As such, I expect major-version bumps to be the norm (as it is currently done).
 * related to the above, I'd like not to have to move directories around for each breaking change (i.e. `/vX` -> `/vY`). I think this can/should be done via the top-level `module` definition.
 * gomod support in libraries consumed here still needs some adjustments, see https://github.com/godbus/dbus/issues/125

**Comment by Xuanwo** [comment]
Maybe we can switch to go mod in next release? 

Use `https://github.com/coreos/go-systemd/v20` as module name just like `https://github.com/google/go-github`.

**Comment by lucab** [comment]
I'd like to do a couple of first-hand experiments before that, in particular wrt. subpackages and module-version, so likely not targeting this at the next release.
And yes, puttting the version label in the `module` directive was my approach in the above list.

**Comment by lucab** [comment]
Also relevant here: go-systemd consumes `coreos/pkg` which is a mix bag of messy thing and not very curated. But, we strictly only need `coreos/pkg/dlopen`, which is quite stable and easier to version.

I'll check and arrange to get the latter moved to `coreos/dlopen` next, with its own semver and go module.

**Comment by lucab** [comment]
Additionally, it is my current understanding that we have to rewrite all our internal imports to be in the form `github.com/coreos/go-systemd/vX/sub_package` and keep the `vX` portion in sync with the `module` declaration.

**Comment by Xuanwo** [comment]
Yes, we need to update all files while upgrading major version.

Golang may add `go mod upgrade` to help reduce the work, but not accepted for now.

**Comment by saschagrunert** [comment]
Hey, any idea how to move forward here? The latest version of godbus/dbus is now using a v5 import suffix which makes it hard to use go-systemd with projects using go modules.

**Comment by lucab** [comment]
@saschagrunert thanks for pinging back. I think we can more or less stick to the plan highlighted so far.
In particular, I think we can proceed as follows:
 * copy `dlopen` under an `internal/` hierarchy here and adjust imports
 * release `v21` with that, which would be the last version using `dep`
 * swap manifests dep->gomod, and introduce `v21` in all imports
 * bump everything to `v22` and release a `v22.0.0` tag

I don't have any of this on my current radar as I'm unfortunately already fully-booked on other topics, but I'll gladly review any PR and try to quickly cut releases when ready. If you have a pressing timeline on this, feel free to start PRing.

**Comment by ulidtko** [comment]
What a coincidence; I'm currently struggling with exactly this issue, too.

I don't feel proficient enough in Go to roll a PR, so the only thing left for me is to wait. Sigh...

File: go.sum [hunk 1]
```diff
@@ -0,0 +1,4 @@
+github.com/godbus/dbus v4.1.0+incompatible h1:WqqLRTsQic3apZUK9qC5sGNfXthmPXzUZ7nQPrNITa4=
+github.com/godbus/dbus v4.1.0+incompatible/go.mod h1:/YcGZj5zSblfDWMMoOzV4fas9FZnQYTkDnsGvmh2Grw=
```
Comment (lucab) [inline go.sum]: This line and the one above look spurious, no?

Comment (saschagrunert) [reply go.sum]: Yes, I will remove it.

**Comment by saschagrunert** [comment]
Would it be fine if we drop the build for go 1.10? Because the v5 imports won't work with that.

**Comment by saschagrunert** [comment]
PTAL

File: .travis.yml [hunk 2]
```diff
@@ -5,14 +5,15 @@ services:
 
 language: go
 go:
-  - "1.10.x"
```
Comment (lucab) [inline .travis.yml]: This is fine, but it the badge in the README needs to be updated as well.

File: go.mod [hunk 3]
```diff
@@ -0,0 +1,5 @@
+module github.com/coreos/go-systemd
```
Comment (lucab) [inline go.mod]: I think we need to introduce the `/v21` here, no?

File: .travis.yml (hunk 2 above)
Comment (saschagrunert) [reply .travis.yml]: Yep, I'll adapt.

File: go.mod (hunk 3 above)
Comment (saschagrunert) [reply go.mod]: Ah yes, I'll change that and adapt the imports.

**Comment by chrboe** [comment]
Is there some kind of workaround for people trying to use this package in a module environment, or will we just have to wait around for this to get merged?

**Comment by saschagrunert** [comment]
> Is there some kind of workaround for people trying to use this package in a module environment, or will we just have to wait around for this to get merged?

I right now stick to an older version.

@lucab do you think we can move forward here?

**Comment by lucab** [comment]
Ah, I didn't realize this was re-pushed after my last review. LGTM now, I'll merge.

**Comment by chrboe** [comment]
Wow, that was quick! Thanks

**Comment by lucab** [comment]
Tagged: https://github.com/coreos/go-systemd/releases/tag/v22.0.0

**Pull request merged**